compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,187 @@
+                                 indicesTy.getElementType(), index);
+                             b.create<linalg::YieldOp>(cast.getResult());
+                           })
+                       .getResult(0);
+      }
+
+      indicesTy = indicesTy.clone(
+          {indicesTy.getDimSize(0), indicesTy.getDimSize(1), 2});
+      indices = builder.create<tosa::ConcatOp>(indicesTy,
+                                               ValueRange{batchIdx, indices},
+                                               rewriter.getI64IntegerAttr(2));
+    }
+
+    auto collapseBatch = [](Value value, ImplicitLocOpBuilder &b) -> Value {
+      auto valueTy = value.getType().cast<ShapedType>();
+      llvm::SmallVector<int64_t> collapseShape(valueTy.getShape().drop_front());
+      llvm::SmallVector<ReassociationExprs, 4> collapseMap(valueTy.getRank() -
+                                                           1);
+      collapseMap.front().push_back(b.getAffineDimExpr(0));
+      for (int i = 0; i < collapseMap.size(); i++) {
+        collapseMap[i].push_back(b.getAffineDimExpr(i + 1));
+      }
+
+      int64_t batch = valueTy.getShape().front();
+      int64_t rows = collapseShape.front();
+      int64_t batchDyn = batch == ShapedType::kDynamic;
+      int64_t rowsDyn = rows == ShapedType::kDynamic;
+      collapseShape[0] =
+          (batchDyn || rowsDyn) ? ShapedType::kDynamic : batch * rows;
+
+      return b.create<tensor::CollapseShapeOp>(valueTy.clone(collapseShape),
+                                               value, collapseMap);
+    };
+
+    indices = collapseBatch(indices, builder);
+    updates = collapseBatch(updates, builder);
+
+    // Create the LinalgExt scatter operation.
+    auto scatter = builder.create<IREE::LinalgExt::ScatterOp>(
+        TypeRange{values.getType()}, ValueRange{updates, indices},
+        ValueRange{values}, builder.getDenseI64ArrayAttr({0, 1}),
+        builder.getBoolAttr(true));
+
+    llvm::SmallVector<Location> argLocs(2, op.getLoc());
+    llvm::SmallVector<Type> args(2, valuesTy.getElementType());
+    Block *scatterBody =
+        builder.createBlock(&scatter.getRegion(), {}, args, argLocs);
+    builder.setInsertionPointToStart(scatterBody);
+    builder.create<IREE::LinalgExt::YieldOp>(scatterBody->getArgument(0));
+    rewriter.replaceOp(op, scatter.getResult(0));
+    return success();
+  }
+};
+
+struct TosaToLinalgExtPass : public TosaToLinalgExtBase<TosaToLinalgExtPass> {
+  void runOnOperation() override {
+    RewritePatternSet patterns(&getContext());
+    ConversionTarget target(getContext());
+    target.addLegalDialect<IREE::LinalgExt::IREELinalgExtDialect,
+                           linalg::LinalgDialect, tensor::TensorDialect,
+                           tosa::TosaDialect>();
+
+    // Not every TOSA op can be legalized to linalg_ext.
+    target.addIllegalOp<tosa::ScatterOp>();
+    target.markUnknownOpDynamicallyLegal([](Operation *) { return true; });
```

Comment:
What if we just removed all of these? (with exception of 168 as you want full conversion to fail if any are left behind - is that true of you used this pass in isolation or only true in the current pipeline?).

Reply:
Yup, that cleaned things up. The DialectConversion framework is still a little confusing for me which may explain the verbose dialect list. 👍